docs/guidance/build/change_history_options.md

```diff
@@ -0,0 +1,11 @@
+# Change History Implementation Options
```

Comment:
On re-reading, I think this page is still quite confusing. I suggest we replace it with the following, which is more closely aligned with the suggestion in https://github.com/open-contracting/standard/issues/1412#issuecomment-925014101. If you agree, we can update the worked examples accordingly.

---

# Change history implementation options

The [releases and records model]() allows publishers to disclose the change history and current state of a contracting process by publishing:

* Updates for changes to the process, known as [releases]().
* The latest values of each field, known as a [compiled release]().
* A version history of each field, known as a [versioned release]().

## Releases or compiled releases

To meet the widest range of user needs the preferred approach is to [publish at least releases and compiled releases](change_history.md).

To publish releases, you need to be able to identify what changed and when in your data sources, for example, by using events or last modified dates. However, it isn't always possible to identify what changed and when, for example because your data sources don't implement versioning, events, good identifiers, or proper controls. In which case, publishing releases can lead to incoherent data due to how the [merge routine](merge_routine.md) handles array entries.

Therefore, if you cannot identify what changed and when in your data sources, you ought to publish [compiled releases only](no_change_history.md).

## Incremental or full updates

If you can publish releases, the preferred approach is to implement a [system architecture](system_architecture.md) that can store and provide access to historic releases. In which case, you can choose to publish either incremental updates or full updates. Incremental updates are releases that contain only new or changed fields and full updates are releases that contain new, changed and unchanged fields.

If your system architecture cannot store and provide access to historic releases, then you need to [publish full updates](full_updates.md). Otherwise, users would not be able to access the values of unchanged fields.

Reply:
It looks good to me. Maybe before the "Releases or compiled releases" section, we can add a sentence:
"As part of your OCDS implementation process, you have two main decisions: whether to have a change history or not and if you have a change history, whether to have incremental updates or full updates."
@jpmckinney, any thoughts?

Reply:
> The heading should be "Individual or compiled releases" rather than "releases or compiled releases".

Sounds good

> That said, as you suggest, the question is more "change history or not", and the conclusion is "individual (and compiled) releases, or only compiled releases".

I think that we might create some confusion by referring to the publication of individual releases without storing and publishing historic releases (i.e. the old 'easy releases' model) as a change history. I know that model means that a 3rd party can maintain a change history if they scrape the data regularly, but from the publisher's point of view it looks more like just publishing the latest version of the data albeit with careful attention to setting ids etc.

I also think that, in terms of the implications for the software they need to write, the decision about whether or not to store and publish historic releases is probably more significant for a publisher than deciding between publishing individual or compiled releases.

Reply:
Good points! So are the decisions then:

* Change history or not (i.e. having multiple releases available at once about the same contracting process, or not)

If no change history, then:

* Individual releases ("easy releases") or compiled releases

If change history, then:

* Incremental updates or full updates

Reply:
Yes, I think that's right. I've taken a pass at reframing along those lines below. I also added some criteria for choosing between incremental and full updates. I think there are a couple of other decisions for implementers that choose to publish a change history, but I think the page gets complicated if we start to include them:

* Whether to publish records (as well as releases)
* If publishing records:
  * Whether to publish compiled releases
  * Whether to publish versioned releases

---

# Change history implementation options

To implement OCDS, you need to decide whether or not to publish a change history.

Using the [releases and records model](), you can publish:

* Updates for changes to the process, known as [releases]().
* The latest values of each field, known as a [compiled release]().
* A version history of each field, known as a [versioned release]().

To publish a change history, you need to implement a [system architecture](system_architecture.md) that can store and provide access to historic releases. You also need to be able to identify what changed and when in your data sources, for example, by using events or last modified dates.

However, it isn't always possible to identify what changed and when, for example, because your data sources don't implement versioning, events, good identifiers, or proper controls. In which case, publishing releases can lead to incoherent data due to how the [merge routine](merge_routine.md) handles array entries.

If you choose to publish a change history, you need to decide whether to publish [incremental or full updates]().

If you choose not to publish a change history, you need to decide whether to publish [individual releases or compiled releases]().

## Incremental or full updates

Incremental updates are releases that contain only new or changed fields and full updates are releases that contain new, changed and unchanged fields.

Incremental updates are smaller than full updates, which can reduce the size of your datastore and reduce download times for data users. If you anticipate many releases per contracting process or very many contracting processes, you ought to consider incremental updates.

If you choose to publish incremental updates, the preferred approach is to also publish complied releases so that users can easily access the latest state of the contracting process.

## Individual releases or compiled releases

If you can identify what changed and when in your data sources, then you ought to [publish individual releases](). To publish individual releases, you need to update the release identifier each time there is a change to the data about a contracting process. That way, data users can easily identify when there is a change. You also need to pay careful attention to consistent use of identifiers for objects in arrays.

If you cannot identify what changed and when, then you ought to [publish compiled releases only]().

Reply:
> I think there are a couple of other decisions for implementers that choose to publish a change history, but I think the page gets complicated if we start to include them:
> 
> * Whether to publish records (as well as releases)
> * If publishing records:
>   
>   * Whether to publish compiled releases
>   * Whether to publish versioned releases

Agreed, let's leave records out for this page.

I've made comments and small edits below.

On word choice: Use "determine" when the meaning is "establish" or "ascertain" in general. Use "identify" only when you are establishing the identity of something or when working with identifiers.

> # Change history implementation options

Since one option is "no change history", this page title could use some work. I don't have a suggestion atm.

> First, you need to decide whether or not to publish a change history.

> Using the releases and records model, you can publish:
> 
> * Updates for changes to the process, known as releases.
> * The latest values of each field, known as a compiled release.
> * A version history of each field, known as a versioned release.

I don't think these bullets serve a purpose, and they break up the logical flow, so we can remove them. If they are here to remind readers of terms' definitions, we can instead describe the terms the first time they are used.

> To publish a change history, you need to be able to determine what changed and when within your data sources: for example, by using event-based triggers or last-modified dates. You also need to use a [system architecture](system_architectures) that has a [separate OCDS datastore](system_architectures.md#separate-ocds-datastore).

The second sentence now follows logically from the first, and I made the second sentence more explicit about the requirement.

> Sometimes, it isn't possible to determine what changed and when: for example, because your data sources don't implement triggers, timestamps, versioning, or consistent identifiers. If identifiers are inconsistent, publishing a change history can lead to incoherent data, due to how the [merge routine](../../schema/merging) handles array entries.

Not sure what "proper controls" implies, so I removed it. If we add it back, we should give an example.

> If you choose to publish a change history, you need to decide whether to publish [incremental or full updates](#incremental-or-full-updates).
> 
> If you choose not to publish a change history, you need to decide whether to publish [individual releases or compiled releases](#individual-releases-or-compiled-releases).
> 
> ## Incremental or full updates
>
> A full update is a [release](../../schema/reference) that contains all the available information about the contracting (or planning) process. An incremental update is a release that contains a subset of the available information, focusing on what's new or changed.

An incremental update isn't *only* new or changed, since it might re-state, for example, the startDate when only the endDate changed. I reworded.

> Incremental updates are smaller than full updates. Smaller releases reduce the OCDS datastore's size and users' download times. If you anticipate many releases per contracting process or very many contracting processes, you ought to consider incremental updates.
> 
> If you choose to publish incremental updates, the preferred approach is to also publish [complied releases](../../schema/records_reference.md#compiled-release), so that users can easily access the latest state of the contracting process.
> 
> ## Individual releases or compiled releases
>
> If you can determine what changed and when within your data sources – but cannot implement a separate OCDS datastore – then you ought to publish individual releases. To publish individual releases, you need to update the release identifier along with each change to the data about a contracting process. That way, users can track each update to the contracting process. You also need to ensure that identifiers are consistent for objects in arrays.

I added a second condition to the first sentence, as otherwise they should implement a change history.

> If you cannot determine what changed and when, then you ought to publish compiled releases only.

Reply:
Those updates look good to me, with one addition to the second sentence below in order to make the relationship between publishing a change history and having a separate OCDS datastore explicit:

> To publish a change history, you need to be able to determine what changed and when within your data sources: for example, by using event-based triggers or last-modified dates. You also need to use a [system architecture](https://github.com/open-contracting/standard/pull/system_architectures) that has a [separate OCDS datastore](https://github.com/open-contracting/standard/pull/system_architectures.md#separate-ocds-datastore) so that you can provide access to historic releases.

Reply:
Sounds good!

Reply:
Resolving as the content has been redrafted.